### ISSUE TITLE
ConcurrentModificationException in AllRulesCache.findInEmfObject()

### ISSUE THREAD

**Issue by ckulla** [eclipse/xtext-core#906]
I see from time to time a `ConcurrentModificationException` while parsing files with Xtext 2.14.0.

```
java.util.ConcurrentModificationException
	at org.eclipse.emf.common.util.ArrayDelegatingEList$EIterator.checkModCount(ArrayDelegatingEList.java:883)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.doNext(AbstractEList.java:699)
	at org.eclipse.emf.common.util.AbstractEList$EIterator.next(AbstractEList.java:685)
	at org.eclipse.xtext.service.AllRulesCache.findInEmfObject(AllRulesCache.java:64)
	at org.eclipse.xtext.parser.antlr.AbstractInternalAntlrParser.createAllRules(AbstractInternalAntlrParser.java:239)
	at org.eclipse.xtext.parser.antlr.AbstractInternalAntlrParser.registerRules(AbstractInternalAntlrParser.java:232)
	at com.<>.parser.antlr.internal.Internal<>Parser.<init>(Internal<>Parser.java:124)
	at com.<>.parser.antlr.<>Parser.createParser(<>Parser.java:25)
	at com.<>.parser.antlr.<>Parser.createParser(<>Parser.java:1)
	at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.doParse(AbstractAntlrParser.java:92)
	at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.parse(AbstractAntlrParser.java:84)
	at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.doParse(AbstractAntlrParser.java:62)
	at org.eclipse.xtext.parser.AbstractParser.parse(AbstractParser.java:33)
	at org.eclipse.xtext.resource.XtextResource.doLoad(XtextResource.java:176)
	at org.eclipse.xtext.linking.lazy.LazyLinkingResource.doLoad(LazyLinkingResource.java:100)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1563)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1342)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoad(ResourceSetImpl.java:259)
	at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper(ResourceSetImpl.java:274)
	at org.eclipse.xtext.resource.XtextResourceSet.getResource(XtextResourceSet.java:265)
```

This happens in a multithreaded application, where files are parsed concurrently on differnet threads.
The application is a kind of background daemon for code generation in a C++ build system.
It happens only sporadically. 

The EMF adapters list of the grammar object seems to get modified on another thread while looking up the `AllRulesCache` instance in `AbstractInternalAntlrParser.createAllRules()`.

**Comment by tivervac** [comment]
AFAIK, EMF is not threadsafe so trouble is to be expected

**Comment by ckulla** [comment]
Yep, still I think my use case should be supported. The application works on different resource sets per thread.

**Comment by cdietrich** [comment]
@ckulla can you please provide a PR

**Comment by ckulla** [comment]
@cdietrich I don't have a solution yet. Any pointers?

**Comment by cdietrich** [comment]
call createParser on a lock?

**Comment by cdietrich** [comment]
@szarnekow  any idea?

**Comment by szarnekow** [comment]
I would try to avoid a synchronized createParser. We could add a boolean value to the `@EmfAdaptable` whether the adapter is to be found in a concurrent env. In that case, we could change the lookup / store logic to be synchronized.

We could even think about installing the given adapter already when the Grammar instance is produced.

**Comment by ckulla** [comment]
@szarnekow  Thanks for the quick answer. I'm still wondering what kind of modification are causing the exception. Are there other places where objects/adapters are added to the grammar object?

**Comment by cdietrich** [comment]
do you use multithreaded resource loading?

**Comment by ckulla** [comment]
@cdietrich  Yes. The application is a headless application (a jetty based http server), where clients send http requests for code generation. Each request leads to resource loading in a separate resource set. The application prevents starting a JVM for each code generation request in our C++ build system. The application receives requests in parallel and processes them in separate threads.

**Comment by cdietrich** [comment]
=> you use a single resourceset in multiple threads?

**Comment by cdietrich** [comment]
=> what about triggering AllRulesCache.findInEmfObject(grammar); directly after creating the standalonesetup.

e.g. injector.get(MyDslGrammarAccess)
AllRulesCache.findInEmfObject(ga.getGrammar)
// pseudo code

**Comment by ckulla** [comment]
> We could even think about installing the given adapter already when the Grammar instance is produced.

Seems the be already the case:

https://github.com/eclipse/xtext-core/blob/5aa53f830144436d05b32fbf5abd4a9e5beb13fa/org.eclipse.xtext/src/org/eclipse/xtext/service/GrammarProvider.java#L74

**Comment by ckulla** [comment]
> => you use a single resourceset in multiple threads?

No

**Comment by cdietrich** [comment]
=> can you try to use a Listener to find out who changes the adapters?

**Comment by ckulla** [comment]
Looks like the AllRulesCache was introduced with 2.14. This matches with my experience, the problem is present since upgrading to 2.14

**Comment by cdietrich** [comment]
still the question: what other adapters are added to the grammar and who does it.

**Comment by cdietrich** [comment]
e.g. do you use serializer?

**Comment by cdietrich** [comment]
@szarnekow do you remember what you had to make threadsafe there in your project?

**Comment by ckulla** [comment]
> e.g. do you use serializer?

Yes

**Comment by cdietrich** [comment]
=> maybe some other caches are added to the grammar in parallel to the read/creation of the parser

**Comment by szarnekow** [comment]
> Seems already be the case:

I suspect the following:

Thread 1: (a) check grammar == null -> (b) go into the sync block -> (c) set this.grammar -> (d) install adapter
Thread 2: starts between (c) and (d) -> grammar is no longer null -> use the object while another thread is still modifying it.

**Comment by szarnekow** [comment]
Potential solution: set this.grammar as the last step in the synchronized block. E.g. extract a method private Grammar doGetGrammar() {..} and use synchronized(this) { if (grammar == null) grammar = doGetGrammar() }

**Comment by ckulla** [comment]
thx a lot, will check this on Monday and provide a PR.

**Comment by ckulla** [comment]
We need to reopen the issue. The issue still occurs.

The grammar object has the following three adapters attached:

```
eAdapters = [
	org.eclipse.xtext.service.AllRulesCache$AllRulesCacheAdapter@3ca175a3, 
	org.eclipse.xtext.serializer.analysis.GrammarElementDeclarationOrder@19ce2fbe, 
	org.eclipse.xtext.xtext.RuleNames$Adapter@2854944e]
```

There seems to be a modification on the adapter list on another thread while looking up the `AllRulesCache` instance in `AbstractInternalAntlrParser.createAllRules()`. The `GrammarElementDeclarationOrder` and `RuleNames$Adapter` are probably added later and interfere then with the `createAllRules()` call.

@szarnekow Is it valid to do multi threaded parsing (with separate resources sets per thread)? Is this a rare use case?

**Comment by cdietrich** [comment]
as i said: the problem is that they may be other adapters attached (lazy) to the grammar.
these should be attached eager before starting multithreading

**Comment by ckulla** [comment]
Eager attaching would mean that the `GrammarProvider` needs to know about the adapters of several other classes. This would increases coupling of the classes. We see this already with the AllRulesCache approach.

Another approach would be to protect lookup and manipulation of the grammar adapters from concurrent read and write. Either with synchronized code blocks or a `ReadWriteLock`.

**Comment by szarnekow** [comment]
I would even prefer a single adapter that maintains a concurrency aware list / map of adapters ... something like a composite adapter. But that would be very involving. I'm with @cdietrich here: Would it be an option for you to init the necessary adapters before you start parsing?

**Comment by szarnekow** [comment]
> Is it valid to do multi threaded parsing (with separate resource sets per thread)? Is this a rare use case?

Yes, this is a valid usecase and happens regularly in other projects, too.

**Comment by ckulla** [comment]
Unfortunately the type `org.eclipse.xtext.xtext.RuleNames.Adapter` is not visible.

**Comment by cdietrich** [comment]
but you can call e.g. org.eclipse.xtext.xtext.RuleNames.getRuleNames(Grammar, boolean) ?

**Comment by cdietrich** [comment]
the main problem: api changes. @szarnekow i dont see how we can fix this without breaking clients

**Comment by ckulla** [comment]
> but you can call e.g. org.eclipse.xtext.xtext.RuleNames.getRuleNames(Grammar, boolean) ?

`new RuleNames(grammar,true)` should do the trick. Will try...

**Comment by ckulla** [comment]
This is what I have now (part of a custom GrammarProvider):

```
@SuppressWarnings("restriction")
private Grammar doGetGrammar(Object requestor) {
    Grammar grammar = super.getGrammar(requestor);
    // install additional adapters to the grammar object in order to prevent
    // ConcurrentModificationException on adapters list of the grammar object
    org.eclipse.xtext.xtext.RuleNames.getRuleNames(grammar, true);
    org.eclipse.xtext.serializer.analysis.GrammarElementDeclarationOrder.get(grammar);
    return grammar;
}
```

Both adapter installation calls are restricted as they are not part of the API. Seems to work so far.
I would prefer @szarnekow's suggestion to have a concurrency aware list / map of adapters.
